batch/test/test_batch.py

```diff
@@ -420,3 +422,35 @@ def test_port(self):
         batch = b.wait()
         print(j.log())
         assert batch['state'] == 'success', batch
+
+    def test_restartable_insert(self):
+        builder = self.client.create_batch()
+        i = 0
+
+        def every_third_time():
+            nonlocal i
+            i += 1
+            if i % 3 == 0:
+                return True
+            return False
+        for i in range(9):
```

Comment:
This rewrite the i above which doesn't really matter but probably not what you intend.  Just a bad code smell.